### PR TITLE
Copy bug

### PR DESCRIPTION
**Pull request by OwenArnold**
Found bug while getting `scippneutron` working with shallow-copy changes:

```python
import scipp as sc
import numpy as np

ds = sc.Dataset(
        data={'a':sc.DataArray(data=sc.Variable(['tof'], values=np.arange(10)),
        coords={'tof': sc.Variable(['tof'], values=np.arange(11))})})
ds.copy()
```
```bash
DimensionError: Inconsistent size for dim 'tof', given 11, requested 10
```

### CODE REVIEW

**Comment by OwenArnold** [comment]
https://github.com/scipp/scipp/blob/main/dataset/dataset.cpp#L120 is still a problem!

File: dataset/operations.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ DataArray copy(const DataArray &array, const AttrPolicy attrPolicy) {
 
 /// Return a deep copy of a Dataset.
 Dataset copy(const Dataset &dataset, const AttrPolicy attrPolicy) {
-  Dataset out({}, copy(dataset.coords()));
+  Dataset out{};
   for (const auto &item : dataset)
     out.setData(item.name(), copy(item, attrPolicy));
```
Comment (SimonHeybrock) [inline dataset/operations.cpp:54]: This will drop coords of datasets without items. Need to insert coords after data?

**Comment by SimonHeybrock** [comment]
> https://github.com/scipp/scipp/blob/main/dataset/dataset.cpp#L120 is still a problem!

What is the problem there?

**Comment by SimonHeybrock** [comment]
Oh, mean it fails to insert items with bin edges, if the dataset previously did not have that dim?

**Comment by SimonHeybrock** [comment]
... but shouldn't that be covered by the `setDims` call https://github.com/scipp/scipp/blob/936ec2d87ac20e156285fab096381dc172d4148b/dataset/dataset.cpp#L115

File: dataset/dataset.cpp [hunk 2]
```diff
@@ -21,6 +21,10 @@ void Dataset::clear() {
   rebuildDims();
 }
 
+void Dataset::setCoords(Coords other) {
+  scipp::expect::contains(other.sizes(), m_coords.sizes());
+  m_coords = other;
```
Comment (SimonHeybrock) [inline dataset/dataset.cpp:26]:
```suggestion
  m_coords = std::move(other);
```

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**